### PR TITLE
REF/BUG: Index.get_value called incorrectly, de-duplicate+simplify

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also engine.get_value is being called incorrectly in DTI/TDI, which this fixes.

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -1915,7 +1915,12 @@ def test_get_value(self, index):
         values = np.random.randn(100)
         value = index[67]
 
-        tm.assert_almost_equal(index.get_value(values, value), values[67])
+        with pytest.raises(AttributeError, match="has no attribute '_values'"):
+            # Index.get_value requires a Series, not an ndarray
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: shouldn't this rather be a TypeError / AssertionError?

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -4618,48 +4617,40 @@ def get_value(self, series, key):
...
 
-        else:
-            if is_scalar(loc):
-                tz = getattr(series.dtype, "tz", None)
-                return libindex.get_value_at(s, loc, tz=tz)
-            return series.iloc[loc]
+        return self._get_values_for_loc(series, loc)
+
+    def _get_values_for_loc(self, series, loc):
+        """
```
Comment (jreback) [inline pandas/core/indexes/base.py]: add Paramters / Returns & typing as much as you can.

This likely duplicates some existing routines, no?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/test_base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py]: only if we add the checks inside get_value.  My thought is that for internal methods with annotated types, we should be able to assume they are called correctly.

Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py]: only if we add the checks inside get_value.  My thought is that for internal methods with annotated types, we should be able to assume they are called correctly.

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: No, it de-duplicates code that we have I think in two places in Index.get_value and one each in DTI and TDI.

That said, this is one of the things we'll be able to simplify if/when we do the _values thing mentioned in #31037.

Will annotate in  next pass

**Review by jreback — APPROVED** [review]

**Pull request merged**